### PR TITLE
Adding overload for DirichletBC

### PR DESCRIPTION
**Pull request by finsberg**
This PR introduces the overloaded `DirichletBC` type and its corresponding PyAdjoint block (`DirichletBCBlock`) to dolfinx-adjoint. This enables the tracing and optimization of boundary conditions that depend on control parameters, such as time-varying boundary values or boundary control problems.

**Key Changes**:

- Added `DirichletBC` overload (`src/dolfinx_adjoint/types/dirichletbc.py`): Wraps dolfinx.fem.DirichletBC as a FloatingType and eagerly annotates it to the working tape.

- Added `DirichletBCBlock` (`src/dolfinx_adjoint/blocks/dirichletbc.py`): A PyAdjoint block that registers the underlying FEniCSx Function or Constant as a dependency and handles the forward recomputation.

- In-place Tape Replay (`src/dolfinx_adjoint/blocks/function_assigner.py`): Updated FunctionAssignBlock to update the FEniCSx function arrays in-place during recompute_component(). 

- Solver Dependency Tracking (`src/dolfinx_adjoint/blocks/solvers.py`): Updated LinearProblemBlock to correctly detect and register dependencies from provided boundary conditions.

- Testing (`tests/test_dirichlet_bc.py`): Added test coverage for tape recording, the annotate=False bypass, block recomputations, and a full time-dependent PDE tape replay.

### CODE REVIEW

File: src/dolfinx_adjoint/blocks/dirichletbc.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from pyadjoint.block import Block
+
+
+class DirichletBCBlock(Block):
+    def __init__(
+        self,
+        value: dolfinx.fem.Function | dolfinx.fem.Constant,
+        dofs: npt.NDArray[np.int32],
+        V: dolfinx.fem.FunctionSpace | None = None,
+        ad_block_tag=None,
```
Comment (jorgensd) [inline src/dolfinx_adjoint/blocks/dirichletbc.py]:
```suggestion
        ad_block_tag: str = None,
```

File: src/dolfinx_adjoint/blocks/dirichletbc.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    ):
+        super().__init__(ad_block_tag=ad_block_tag)
+        self.dofs = dofs
+        self.V = V
```
Comment (jorgensd) [inline src/dolfinx_adjoint/blocks/dirichletbc.py]:
```
        self.dofs = dofs
        self.V = V
```
Private with property?

File: src/dolfinx_adjoint/blocks/solvers.py [hunk 3]
```diff
@@ -354,8 +352,23 @@ def evaluate_tlm_component(self, inputs, tlm_inputs, block_variable, idx, prepar
...
+            self._ksp_tlm.setOperators(self._A_tlm)
+
+            prefix = self._petsc_options_prefix + "tlm_"
+            self._ksp_tlm.setOptionsPrefix(prefix)
+            if self._tlm_petsc_options is not None:
+                opts = PETSc.Options()
+                for k, v in self._tlm_petsc_options.items():
+                    opts[prefix + k] = v
+            self._ksp_tlm.setFromOptions()
+            self._ksp_tlm.setUp()
```
Comment (jorgensd) [inline src/dolfinx_adjoint/blocks/solvers.py]:
```suggestion
        # Create and assemble TLM matrix
        if not hasattr(self, "_A_tlm"):
            self._A_tlm = dolfinx.fem.petsc.create_matrix(dFdu)
        self._A_tlm.zeroEntries()
        dolfinx.fem.petsc.assemble_matrix(self._A_tlm, dFdu, bcs=bcs)
        self._A_tlm.assemble()
        
        # Create TLM KSP and attach matrix
        if not hasattr(self, "_ksp_tlm")
            self._ksp_tlm = PETSc.KSP().create(self._A_tlm.getComm())
        self._ksp_tlm.setOperators(self._A_tlm)
  
        # Set TLM solver options
        if self._tlm_petsc_options is not None:
            prefix = self._petsc_options_prefix + "tlm_"
            self._ksp_tlm.setOptionsPrefix(prefix)
            opts = PETSc.Options()
            opts.prefixPush(solver_prefix)
            for k, v in solver_parameters.items():
                opts.setValue(k, v)
            self._ksp_tlm.setFromOptions()
            opts.prefixPop()
   
            # For some strange reason delValue doesn't respect prefixes
            for k, v in solver_parameters.items():
                opts.delValue(f"{solver_prefix}{k}")
        # Setup preconditioner
        self._ksp_tlm.setUp()

```

File: src/dolfinx_adjoint/types/dirichletbc.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import numpy.typing as npt
+import pyadjoint
+from pyadjoint.overloaded_type import FloatingType
+
+from ..blocks.dirichletbc import DirichletBCBlock
+from .function import Function
+
+
+class DirichletBC(dolfinx.fem.DirichletBC, FloatingType):
+    """A class overloading `dolfinx.fem.DirichletBC` to support it being used as a control variable
```
Comment (jorgensd) [inline src/dolfinx_adjoint/types/dirichletbc.py]:
```suggestion
    """A class overloading :py:class:`dolfinx.fem.DirichletBC` to support it being used as a control variable
```

File: src/dolfinx_adjoint/types/dirichletbc.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    in the adjoint framework.
+
+    Args:
+        g: The value of the Dirichlet BC.
+        dofs: An array of degree-of-freedom indices in `V` where the BC should be applied.
+        **kwargs: Additional keyword arguments to pass to the `pyadjoint.overloaded_type.FloatingType` constructor.
```
Comment (jorgensd) [inline src/dolfinx_adjoint/types/dirichletbc.py]:
```suggestion
        **kwargs: Additional keyword arguments to pass to the :py:func:`pyadjoint.overloaded_type.FloatingType` constructor.
```

File: src/dolfinx_adjoint/types/dirichletbc.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            bctype = dolfinx.cpp.fem.DirichletBC_float64
+        elif np.issubdtype(dtype, np.complex64):
+            bctype = dolfinx.cpp.fem.DirichletBC_complex64
+        elif np.issubdtype(dtype, np.complex128):
+            bctype = dolfinx.cpp.fem.DirichletBC_complex128
+        else:
+            raise NotImplementedError(f"Type {dtype} not supported.")
+
+        bc_kwargs = {}
+        # If dolfinx-version is 0.12 we need to pass the following
```
Comment (jorgensd) [inline src/dolfinx_adjoint/types/dirichletbc.py:51]: Use packaging version here so that we only have a single super().__init__ below.

File: src/dolfinx_adjoint/types/dirichletbc.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            super().__init__(bctype(g._cpp_object, dofs))
+
+        annotate = kwargs.pop("annotate", True)
+        annotate = annotate and pyadjoint.annotate_tape()
+
+        FloatingType.__init__(
+            self,
+            g,
+            dtype=dtype,
+            block_class=kwargs.pop("block_class", DirichletBCBlock),
```
Comment (jorgensd) [inline src/dolfinx_adjoint/types/dirichletbc.py:66]: Skip kwargs as it is unclear to me what should happen if a user sends in a different block.

Comment (jorgensd) [reply src/dolfinx_adjoint/types/dirichletbc.py:66]: Keep to make this modular.

File: src/dolfinx_adjoint/types/dirichletbc.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    def _ad_create_checkpoint(self):
+        return self
+
+    def _ad_restore_at_checkpoint(self, checkpoint):
+        return self
+
+
+def dirichletbc(value: Function, dofs: npt.NDArray[np.int32], **kwargs) -> DirichletBC:
+    """Overloaded DirichletBC constructor that creates an adjoint-aware DirichletBC"""
```
Comment (jorgensd) [inline src/dolfinx_adjoint/types/dirichletbc.py]: Maybe expand docs a bit here, making it clear that in DOLFINx-adjoint the input should always be a function, and not a constant?
Also mention the different kwargs based on version?

File: tests/test_dirichlet_bc.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    V = dolfinx.fem.functionspace(mesh, ("Lagrange", 1))
+
+    c = Function(V, name="boundary_value")
+    c.interpolate(lambda x: np.full_like(x[0], 5.0))
+
+    dofs = dolfinx.fem.locate_dofs_geometrical(V, lambda x: np.isclose(x[0], 0.0))
+    bc = dirichletbc(c, dofs)
+
+    tape = pyadjoint.get_working_tape()
+    block = tape.get_blocks()[0]
```
Comment (jorgensd) [inline tests/test_dirichlet_bc.py:75]: We should check that this is a DirichletBCBlock.

File: src/dolfinx_adjoint/types/dirichletbc.py (hunk 8 above)
Comment (jorgensd) [reply src/dolfinx_adjoint/types/dirichletbc.py]: Assert type of `value` and give clear error message stating that they should use da-function or da-constant.

File: src/dolfinx_adjoint/blocks/dirichletbc.py (hunk 2 above)
Comment (finsberg) [reply src/dolfinx_adjoint/blocks/dirichletbc.py]: Fixed in https://github.com/scientificcomputing/dolfinx-adjoint/pull/49/commits/bf7e5de5c887ce1e0e9a54d9c428b68e4962cc05

File: src/dolfinx_adjoint/types/dirichletbc.py (hunk 6 above)
Comment (finsberg) [reply src/dolfinx_adjoint/types/dirichletbc.py:51]: Fixed in https://github.com/scientificcomputing/dolfinx-adjoint/pull/49/commits/a84dfa675a3a5dfa1bcc968c7fa0b01e8debfc4b

File: src/dolfinx_adjoint/types/dirichletbc.py (hunk 8 above)
Comment (finsberg) [reply src/dolfinx_adjoint/types/dirichletbc.py]: Fixed in https://github.com/scientificcomputing/dolfinx-adjoint/pull/49/commits/63950904913957769b4c960d64c2a6b9914e528a

File: tests/test_dirichlet_bc.py (hunk 9 above)
Comment (finsberg) [reply tests/test_dirichlet_bc.py:75]: Fixed in https://github.com/scientificcomputing/dolfinx-adjoint/pull/49/commits/46c796c41ee16dcbc0b379613d106b834060a9fd
